### PR TITLE
Bring back `retryAfter` functionality

### PR DESCRIPTION
**Pull request by fnwbr**
This is in progress, @mea36 is going to complete the code including automated tests + the full description including test and QA steps

---

Closes https://github.com/Expensify/Bedrock/issues/111
Also related to https://github.com/Expensify/Expensify/issues/53817

Original PR was here: https://github.com/Expensify/Bedrock/pull/169
Which was reverted here: https://github.com/Expensify/Bedrock/pull/194

---
Update from @mea36 
I made the following changes to the original PR:

- I added tests for the retryAfter functionality (https://github.com/Expensify/Bedrock/pull/243/commits/0f549c344941fa20069413e6482b450f73493aa5)
- I updated `GetJob` in process to order by `nextRun` **and** `jobID` because two jobs that have the same `nextRun` time should return the job that was inserted first (fix is in 19983f1ffd2d8781536a18cce04cffafe12525df)
- I updated `GetJob` in peek to look for jobs in states `QUEUED` **and** `RUNQUEUED` (fix is in f3ea4f78fdc5b8e5314480284073e9c154c1d520)

### CODE REVIEW

**Comment by mea36** [comment]
I'm going on vacation, so if someone wants to pick this up, feel free to do so.

Update:
I added all the tests from https://github.com/Expensify/Bedrock/pull/195 and https://github.com/Expensify/Bedrock/pull/169
I added these in a quick and dirty way, so I don't think the test names are the best and will change it later

Tests still need to be added from https://github.com/Expensify/Bedrock/pull/191

I had some difficulty after adding the tests for https://github.com/Expensify/Bedrock/pull/169
The test `retryWithChildren` works when run by itself but segfaults when run with all the other tests in the file.  I haven't figured out why

`retryJobComesFirst` fails.  This tests the actual lifecycle of retrying a job, therefore I think the code is buggy.
`retryJobComesFirst` creates 2 retryable jobs and one nonretryable job.  We use `GetJob` to get the first retryable job and then sleep for the retry after value.  After that `GetJob` should (if I understand the functionality correctly - I admittedly haven't looked at the code yet) return the first retryable job.  Instead it gets the nonretryable job.

To run just the CreateJob test execute
```
./test -only CreateJob
```

When `retryWithChildren` fails, it'll keep the process running. That looks like
```
$ ps aux | grep -i bedrock
root     17395  0.0  0.6 260864 13252 ?        Ssl  21:34   0:00 bedrock -fork -db /vagrant/data/auth/main.db -serverHost 0.0.0.0:4444 -nodeName auth-on-bedrock -nodeHost localhost:4445 -keyHost localhost:4001 -priority 200 -v -cache 10001 -debugKeys -plugins db,auth
```
If you don't kill that process then the next time you run the tests they'll fail

**Comment by fnwbr** [comment]
Just to clarify: I'm not working on this; I opened this PR to kickstart this and give someone a base to get started off

By now there are conflicts already again; initially the only thing that would've been to do here was writing tests + testing

**Comment by cead22** [comment]
@fnwbr @mea36 can you guys take ownership of this please since we need it for SmartScan Dos?

**Comment by mea36** [comment]
I assumed this was on hold due to our feature freeze.  If that's the not the case then I can start working on it next week

**Comment by cead22** [comment]
Awesome thank you, and to clarify, given that this is part of SmartScan Dos it should not be on hold.

**Comment by mea36** [comment]
Note to self, I need to add the tests from https://github.com/Expensify/Bedrock/pull/240 as well

**Comment by iwiznia** [comment]
This has no issue attached, thus can't add priority labels. This [should be treated as `Daily` now](https://expensify.slack.com/archives/C6V5B2KTR/p1504022505000063?thread_ts=1504017974.000384&cid=C6V5B2KTR).

**Comment by fnwbr** [comment]
Wrong, the issue is linked above in the initial description. Can't add the `Daily` label though because it's in the Bedrock repo...

**Comment by mea36** [comment]
All tests from #195 and #169 are added.
I didn't like the tests from https://github.com/Expensify/Bedrock/pull/191 so I didn't copy them.  Looking at the comments it looks like @pecanoro created `CancelJob` in one way and then updated the code but didn't update the tests she used to test it. @pecanoro is that correct?
For `CancelJob` (https://github.com/Expensify/Bedrock/pull/191) I instead created tests based on the code.

I also added a test for https://github.com/Expensify/Bedrock/pull/240

There were some bug fixes I had to make
- `CancelJob` was broken for when the jobID did not exist (fix is a83fad52b96e71810bfe1f1176d66a38266cbb39)
- I updated `GetJob` in process to order by `nextRun` **and** `jobID` because two jobs that have the same `nextRun` time should return the job that was inserted first (fix is in 19983f1ffd2d8781536a18cce04cffafe12525df)
- I updated `GetJob` in peek to look for jobs in states `QUEUED` **and** `RUNQUEUED` (fix is in f3ea4f78fdc5b8e5314480284073e9c154c1d520)

Note, all I did was add tests.  I didn't touch the functionality of `retryAfter` if it wasn't to fix a test.  I have some questions about the implementation and will comment on the appropriate lines to start a conversation

**Comment by mea36** [comment]
@flodnv ready for review

**Comment by mea36** [comment]
I'll resolve the conflicts tomorrow.  I'm also going to remove the tests that are unrelated to `retryAfter` so they can be reviewed independently of this functionality

**Comment by mea36** [comment]
Holding on https://github.com/Expensify/Bedrock/pull/275

**Comment by pecanoro** [comment]
> @pecanoro created CancelJob in one way and then updated the code but didn't update the tests she used to test it. @pecanoro is that correct?

Those test should be working, I kept updating them, @cead22 used them to test it before merging the code. What was the problem?

**Comment by iwiznia** [comment]
SSDos first version is almost ready and we'll need this. @mea36 is it close to be done? Do you have an ETA?

**Comment by mea36** [comment]
ETA - EOD today

**Comment by mea36** [comment]
We had too many different versions of the test files to extracted the tests that @cead22 requested I make into https://github.com/Expensify/Bedrock/pull/275 (which was merged and this branch is up-to-date with master)

This is now ready to be reviewed.
- I added tests for the retryAfter functionality (https://github.com/Expensify/Bedrock/pull/243/commits/0f549c344941fa20069413e6482b450f73493aa5)
- I updated `GetJob` in process to order by `nextRun` **and** `jobID` because two jobs that have the same `nextRun` time should return the job that was inserted first (fix is in 19983f1ffd2d8781536a18cce04cffafe12525df)
- I updated `GetJob` in peek to look for jobs in states `QUEUED` **and** `RUNQUEUED` (fix is in f3ea4f78fdc5b8e5314480284073e9c154c1d520)

I'll update the description to reflect this comment

**Comment by iwiznia** [comment]
@pecanoro, @tylerkaraszewski can you review the PR?

**Comment by iwiznia** [comment]
Also, I'm not finding what made us revert this change last time... are we sure it will not happen again?

File: plugins/Jobs.cpp [hunk 1]
```diff
@@ -647,9 +698,35 @@ bool BedrockPlugin_Jobs::processCommand(SQLite& db, BedrockCommand& command) {
...
-            throw "502 Update failed";
+
+        // Update jobs without retryAfter
+        if (!nonRetriableJobs.empty()) {
+            string jobIDs = SComposeList(nonRetriableJobs);
+            SINFO("Updating jobs without retryAfter " << jobIDs);
+            string updateQuery = "UPDATE jobs "
+                                 "SET state='RUNNING', "
+                                 "lastRun=" + SCURRENT_TIMESTAMP() + " "
+                                 "WHERE jobID IN (" + jobIDs + ");";
```
Comment (pecanoro) [inline plugins/Jobs.cpp]: `SQList(nonRetriableJobs)`

File: plugins/Jobs.cpp [hunk 2]
```diff
@@ -647,9 +698,35 @@ bool BedrockPlugin_Jobs::processCommand(SQLite& db, BedrockCommand& command) {
...
-        if (!db.write(updateQuery)) {
-            throw "502 Update failed";
+
+        // Update jobs without retryAfter
+        if (!nonRetriableJobs.empty()) {
+            string jobIDs = SComposeList(nonRetriableJobs);
+            SINFO("Updating jobs without retryAfter " << jobIDs);
+            string updateQuery = "UPDATE jobs "
+                                 "SET state='RUNNING', "
+                                 "lastRun=" + SCURRENT_TIMESTAMP() + " "
```
Comment (pecanoro) [inline plugins/Jobs.cpp]: You are missing the indentation here, right?

File: plugins/Jobs.cpp [hunk 3]
```diff
@@ -647,9 +698,35 @@ bool BedrockPlugin_Jobs::processCommand(SQLite& db, BedrockCommand& command) {
...
+        if (!retriableJobs.empty()) {
+            SINFO("Updating jobs with retryAfter");
+            for (auto job : retriableJobs) {
+                string jobID = job["jobID"];
+                string retryAfter = job["retryAfter"];
+                string updateQuery = "UPDATE jobs "
+                                     "SET state='RUNQUEUED', "
+                                     "  lastRun=" + SCURRENT_TIMESTAMP() + ", "
+                                     "  nextRun=DATETIME(" + SCURRENT_TIMESTAMP() + ", " + SQ(retryAfter) + ") "
+                                     "WHERE jobID = " + jobID + ";";
```
Comment (pecanoro) [inline plugins/Jobs.cpp]: `SQ(jobID)`?

File: plugins/Jobs.cpp [hunk 4]
```diff
@@ -647,9 +698,35 @@ bool BedrockPlugin_Jobs::processCommand(SQLite& db, BedrockCommand& command) {
...
+
+        // Update jobs with retryAfter
+        if (!retriableJobs.empty()) {
+            SINFO("Updating jobs with retryAfter");
+            for (auto job : retriableJobs) {
+                string jobID = job["jobID"];
+                string retryAfter = job["retryAfter"];
+                string updateQuery = "UPDATE jobs "
+                                     "SET state='RUNQUEUED', "
+                                     "  lastRun=" + SCURRENT_TIMESTAMP() + ", "
```
Comment (pecanoro) [inline plugins/Jobs.cpp]: Those spaces `"  last` are weird, you should indent the line instead

File: plugins/Jobs.cpp [hunk 5]
```diff
@@ -704,6 +781,11 @@ bool BedrockPlugin_Jobs::processCommand(SQLite& db, BedrockCommand& command) {
 
         const string& nextRun = result[0][1];
         const string& lastRun = result[0][2];
+        const string& state = result[0][3];
+
+        if (state == "RUNQUEUED") {
+            throw "402 Auto-retrying jobs cannot be updated once running";
```
Comment (pecanoro) [inline plugins/Jobs.cpp]: I am curious, why can't it be updated?

File: plugins/Jobs.cpp [hunk 6]
```diff
@@ -1026,7 +1109,11 @@ string BedrockPlugin_Jobs::_constructNextRunDATETIME(const string& lastScheduled
     }
 
     // Validate the sqlite date modifiers
-    // See: https://www.sqlite.org/lang_datefunc.html
+    if (!_isValidSQLiteDateModifier(SComposeList(parts))){
+        SWARN("Syntax error, failed parsing repeat");
+        return "";
+    }
+
     for (const string& part : parts) {
```
Comment (pecanoro) [inline plugins/Jobs.cpp]: If you created `_isValidSQLiteDateModifier`, this seems unneeded and doing the same thing.

File: plugins/Jobs.cpp (hunk 5 above)
Comment (iwiznia) [reply plugins/Jobs.cpp]: +1, we can update regular running jobs, why wouldn't we be able to do so for ones with retryAfter?

File: plugins/Jobs.cpp [hunk 7]
```diff
@@ -406,6 +432,12 @@ bool BedrockPlugin_Jobs::processCommand(SQLite& db, BedrockCommand& command) {
                 if (!result.empty()) {
                     updateJobID = SToInt64(result[0][0]);
                 }
+
+                // Alert if job is unique and retryAfter is set
+                // This shouldn't happen since we validate this in peekCommand
+                if (SContains(job, "retryAfter") && job["retryAfter"] != "") {
```
[the remark this answers is not in the document]
Comment (mea36) [reply plugins/Jobs.cpp]: This was here from the original code version of this code (remember that we cherry-picked the commits from the original attempt into this branch).

But I agree with you, it's not necessary.  @cead22 is there a specific reason you had this here?

File: plugins/Jobs.cpp [hunk 8]
```diff
@@ -494,8 +532,11 @@ bool BedrockPlugin_Jobs::processCommand(SQLite& db, BedrockCommand& command) {
                     }
                 }
 
+                // If no data was provided, use an empty object
+                const string& safeRetryAfter = job["retryAfter"].empty() ? "\"\"" : SQ(job["retryAfter"]);
```
[the remark this answers is not in the document]
Comment (mea36) [reply plugins/Jobs.cpp]: Actually, `job` is a `STable` which is a `map` which correctly handles a key that doesn't exist http://www.cplusplus.com/reference/map/map/operator[]/

I will change the `"\"\""` to `SQ("")`

[the remark this answers is not in the document]
Comment (pecanoro) [reply plugins/Jobs.cpp]: > Actually, job is a STable which is a map which correctly handles a key that doesn't exist

Ok, perfect then, can you make sure though? So we don't get any weird exceptions

[the remark this answers is not in the document]
Comment (pecanoro) [reply plugins/Jobs.cpp]: I mean, I am more concerned about calling `.empty()` on whatever it returns when the key doesn't exist, which I am not sure if it's a string

File: plugins/Jobs.cpp [hunk 9]
```diff
@@ -601,13 +643,22 @@ bool BedrockPlugin_Jobs::processCommand(SQLite& db, BedrockCommand& command) {
...
+                nonRetriableJobs.push_back(result[c][0]);
+            } else {
+                STable job;
+                job["jobID"] = result[c][0];
+                job["retryAfter"] = result[c][4];
+                retriableJobs.push_back(job);
+            }
 
             // See if this job has any FINISHED/CANCELLED child jobs, indicating it is being resumed
             SQResult childJobs;
```
[the remark this answers is not in the document]
Comment (mea36) [reply plugins/Jobs.cpp]: Good catch

File: plugins/Jobs.cpp (hunk 5 above)
Comment (mea36) [reply plugins/Jobs.cpp]: I have no idea.   @cead22 was this part of the original requirement?

**Comment by mea36** [comment]
I made a few updates.  I think I should make some automated tests for `GetJob` to make sure we're not breaking any functionality. I'll get to that tomorrow (unless someone wants to do that for me :P )

File: test/tests/jobs/CreateJobTest.cpp [hunk 10]
```diff
@@ -234,4 +240,167 @@ struct CreateJobTest : tpunit::TestFixture {
...
+        ASSERT_EQUAL(response["name"], jobName);
+
+        // Query the db and confirm that state, nextRun and lastRun are correct
+        SQResult jobData;
+        tester->readDB("SELECT state, nextRun, lastRun FROM jobs WHERE jobID = " + jobID + ";", jobData);
+
+        ASSERT_EQUAL(jobData[0][0], "RUNQUEUED");
+        string nextRun = jobData[0][1];
+        string lastRun = jobData[0][2];
+        struct tm tm1;
```
Comment (cead22) [inline test/tests/jobs/CreateJobTest.cpp]: Can you please explain this in a comment, specifically this `ASSERT_EQUAL(difftime(nextRunTime, lastRunTime), 1);`?

File: test/tests/jobs/CreateJobTest.cpp [hunk 11]
```diff
@@ -234,4 +240,167 @@ struct CreateJobTest : tpunit::TestFixture {
...
+        struct tm tm2;
+        strptime(nextRun.c_str(), "%Y-%m-%d %H:%M:%S", &tm1);
+        time_t nextRunTime = mktime(&tm1);
+        strptime(lastRun.c_str(), "%Y-%m-%d %H:%M:%S", &tm2);
+        time_t lastRunTime = mktime(&tm2);
+        ASSERT_EQUAL(difftime(nextRunTime, lastRunTime), 1);
+
+        // Get the job, confirm error
+        tester->executeWaitVerifyContent(command, "404 No job found");
+        // Wait 5 seconds, get the job, confirm no error
```
Comment (cead22) [inline test/tests/jobs/CreateJobTest.cpp]: - Blank line before comment
- Comments says 5 seconds but sleep is called with 1

File: test/tests/jobs/FinishJobTest.cpp [hunk 12]
```diff
@@ -353,4 +354,26 @@ struct FinishJobTest : tpunit::TestFixture {
...
+        command["name"] = "job";
+        command["retryAfter"] = "+1 SECOND";
+        STable response = tester->executeWaitVerifyContentTable(command);
+        string jobID = response["jobID"];
+
+        // Get the job
+        command.clear();
+        command.methodLine = "GetJob";
+        command["name"] = "job";
+        tester->executeWaitVerifyContent(command);
```
Comment (cead22) [inline test/tests/jobs/FinishJobTest.cpp]: In previous tests you're asserting that you're getting the right info in the response, why not here?

File: test/tests/jobs/FinishJobTest.cpp [hunk 13]
```diff
@@ -353,4 +354,26 @@ struct FinishJobTest : tpunit::TestFixture {
...
+
+        // Finish it
+        command.clear();
+        command.methodLine = "FinishJob";
+        command["jobID"] = jobID;
+        tester->executeWaitVerifyContent(command);
```
Comment (cead22) [inline test/tests/jobs/FinishJobTest.cpp]: No assertions?

File: test/tests/jobs/RetryJobTest.cpp [hunk 14]
```diff
@@ -278,4 +279,26 @@ struct RetryJobTest : tpunit::TestFixture {
...
+        command.clear();
+        command.methodLine = "GetJob";
+        command["name"] = "job";
+        tester->executeWaitVerifyContent(command);
+
+        // Finish it
+        command.clear();
+        command.methodLine = "RetryJob";
+        command["jobID"] = jobID;
+        tester->executeWaitVerifyContent(command);
```
Comment (cead22) [inline test/tests/jobs/RetryJobTest.cpp]: No assertions?

File: plugins/Jobs.cpp (hunk 5 above)
Comment (pecanoro) [reply plugins/Jobs.cpp]: Yeah, but why is that? Why is it not safe for a RUNQUEUED job but it is for a RUNNING one? Maybe @quinthar knows the answer for this better

File: test/tests/jobs/FinishJobTest.cpp (hunk 12 above)
Comment (mea36) [reply test/tests/jobs/FinishJobTest.cpp]: This is a test for `FinishJob` functionality.  A test for `GetJob` should verify the return values of `GetJob`

File: plugins/Jobs.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (mea36) [reply plugins/Jobs.cpp]: I'm going to take this out.  If this is happening then it's a symptom of a much larger problem (ie `peek` and `process` aren't being run sequentially on master)

File: test/tests/jobs/FinishJobTest.cpp (hunk 13 above)
Comment (mea36) [reply test/tests/jobs/FinishJobTest.cpp]: Added some

File: plugins/Jobs.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (mea36) [reply plugins/Jobs.cpp]: good catch

File: plugins/Jobs.cpp [hunk 15]
```diff
@@ -550,35 +592,35 @@ bool BedrockPlugin_Jobs::processCommand(SQLite& db, BedrockCommand& command) {
...
-                    "SELECT jobID, name, data, priority, parentJobID "
+                    "SELECT jobID, name, data, priority, parentJobID, retryAfter "
                     "FROM jobs "
-                    "WHERE state='QUEUED' "
+                    "WHERE state IN ('QUEUED', 'RUNQUEUED') "
                     "  AND priority=1000"
                     "  AND " + SCURRENT_TIMESTAMP() + ">=nextRun "
                     "  AND name GLOB " + SQ(name) + " "
-                    "ORDER BY nextRun ASC LIMIT " + safeNumResults +
+                    "ORDER BY nextRun, jobID ASC LIMIT " + safeNumResults +
```
[the remark this answers is not in the document]
Comment (mea36) [reply plugins/Jobs.cpp]: 1.  I think what you're asking here is if `ORDER BY priority DESC` does a stable sort.  Is that correct?  If so, I created tests for `GetJob` that test the return order of jobs with the same `priority` and `nextRun` but different  `state` values.  Therefore I'm confident that `order by nextRun, jobID` isn't affected by `ORDER BY priority DESC`.  Check out the `testPrioritiesWithRunQueued` test
1.  I haven't benchmarked this but I did check that the query plan is the same.  If you want, I can benchmark it tomorrow
Query plan without `jobID ASC`
```
~$ sudo readdb.sh "explain query plan SELECT jobID, name, data, priority, parentJobID, retryAfter
> FROM jobs
> WHERE state IN ('QUEUED', 'RUNQUEUED')
>   AND priority=0
>   AND current_timestamp >=nextRun
>   AND name GLOB 'www-prod/*'
> ORDER BY nextRun ASC LIMIT 10;"
selectid    order       from        detail
----------  ----------  ----------  -------------------------------------------------------------------------------------------------
0           0           0           SEARCH TABLE jobs USING INDEX jobsStatePriorityNextRunName (state=? AND priority=? AND nextRun<?)
0           0           0           EXECUTE LIST SUBQUERY 1
0           0           0           USE TEMP B-TREE FOR ORDER BY
```
Query plan with `jobID ASC`
```
~$ sudo readdb.sh "explain query plan SELECT jobID, name, data, priority, parentJobID, retryAfter
> FROM jobs
> WHERE state IN ('QUEUED', 'RUNQUEUED')
>   AND priority=0
>   AND current_timestamp >=nextRun
>   AND name GLOB 'www-prod/*'
> ORDER BY nextRun ASC, jobID ASC LIMIT 10;"
selectid    order       from        detail
----------  ----------  ----------  -------------------------------------------------------------------------------------------------
0           0           0           SEARCH TABLE jobs USING INDEX jobsStatePriorityNextRunName (state=? AND priority=? AND nextRun<?)
0           0           0           EXECUTE LIST SUBQUERY 1
0           0           0           USE TEMP B-TREE FOR ORDER BY
```

File: plugins/Jobs.cpp (hunk 5 above)
Comment (mea36) [reply plugins/Jobs.cpp]: Per https://github.com/Expensify/Bedrock/issues/111#issuecomment-327648440, I'll remove this and create a test for it tomorrow

Comment (mea36) [reply plugins/Jobs.cpp]: Actually, with this removed there is no change to `UpdateJob` so I won't create a test

File: plugins/Jobs.cpp (hunk 15 above)
[the remark this answers is not in the document]
Comment (cead22) [reply plugins/Jobs.cpp]: > created tests for GetJob that test the return order of jobs with the same priority and nextRun but different state values. Therefore I'm confident that order by nextRun, jobID isn't affected by ORDER BY priority DESC

The test should be with the same priority and nextRun but different priorities not different states, right?

What I'm thinking with this is that we _may_ be slowing down one of the main queries to solve what seems to me like a non-isssue which is the fact that Job B which was created after Job A may be processed before Job A if they both have the same priority and the very same nextRun. The reason I call it a non-issue is because I understand the case but it definitely seems like an edge case and I haven't seen that this is causing a problem, am I missing something? Would you agree that this change might not be worth implementing now?

[the remark this answers is not in the document]
Comment (mea36) [reply plugins/Jobs.cpp]: If we generally consider this a non-issue, then I can revert this change.  The reason I implemented it was because this test failed
https://github.com/Expensify/Bedrock/pull/243/files#diff-c8ebb7cec8b361355acfa8a63051dacbR364

Now the following jobs also tests will also fail 
https://github.com/Expensify/Bedrock/pull/243/files#diff-c8ebb7cec8b361355acfa8a63051dacbR268
https://github.com/Expensify/Bedrock/pull/243/files#diff-cf845d40b06e01293a549aa13ac565adR326

So if this is a non-issue, I'll update the tests accordingly

**Comment by iwiznia** [comment]
@mea36 tests are failing... 
Also, @tylerkaraszewski you are listed as reviewer, so can you review this or remove yourself as a reviewer?

**Comment by pecanoro** [comment]
@mea36 forgot to remove one of the tests after the removing https://github.com/Expensify/Bedrock/pull/243/commits/cafa9fcd6cdd78847eccde605846bd0f491c1243

File: plugins/Jobs.cpp (hunk 15 above)
[the remark this answers is not in the document]
Comment (mea36) [reply plugins/Jobs.cpp]: Actually, @quinthar do you have an opinion here? 
I agree with @cead22 that this is an edge case but it's also the proper way to achieve the behavior we want.
If we move forward with this change, I should devise a plan to test this query

[the remark this answers is not in the document]
Comment (quinthar) [reply plugins/Jobs.cpp]: I'm not sure I follow.  Are you asking if it's worth processing same-priority jobs in order, or if it's ok if same-priority jobs execute out of order?  I'd say it would be great to process them in order if possible, and it sounds like @mea36's test confirms that it is not just possible, but has no effect on the query plan (and thus should have the same performance).  Am I misunderstanding something?

[the remark this answers is not in the document]
Comment (cead22) [reply plugins/Jobs.cpp]: No that's exactly right and for added context:
- I agree that Mel's implementation is correct
- I don't think that 2 equal query plans translates into two equal query execution times. My worry here specifically is that the temp btree used to sort could be bigger since it's now using 2 values and thus take longer to create and traverse (which we do 3 times)
- In light of that and our recent query perf degradation problems, I'd love for us to bechmark the new query vs the old one before releasing to be on the safer side
- And because of that I proposed that as an alternative we don't solve this edge case now given that this doesn't solve any problems from the user's perspective, and live with the fact that 2 jobs with the exact same nextRun and the exact same priority maybe be dispatched our of order.

[the remark this answers is not in the document]
Comment (mea36) [reply plugins/Jobs.cpp]: Since the hard work was already done to do some performance testing (thanks @cead22 ), I followed the steps in https://github.com/Expensify/Expensify/issues/49378#issuecomment-298134063 using a jobs table from @cead22 that has 347K jobs;

I created a new database and created a table using the schema I have on dev (minus the `retryAfter` column
```sql
sqlite> .schema jobs
CREATE TABLE jobs ( created TIMESTAMP NOT NULL, jobID INTEGER NOT NULL PRIMARY KEY AUTOINCREMENT, state TEXT NOT NULL, name TEXT NOT NULL, nextRun TIMESTAMP NOT NULL, lastRun TIMESTAMP, repeat TEXT NOT NULL, data TEXT NOT NULL, priority INTEGER NOT NULL DEFAULT 500, parentJobID INTEGER NOT NULL DEFAULT 0);
CREATE INDEX jobsState    ON jobs ( state    );
CREATE INDEX jobsName     ON jobs ( name     );
CREATE INDEX jobsNextRun  ON jobs ( nextRun  );
CREATE INDEX jobsLastRun  ON jobs ( lastRun  );
CREATE INDEX jobsPriority ON jobs ( priority );
CREATE INDEX jobsParentJobIDState ON jobs ( parentJobID, state );
CREATE INDEX jobsStatePriorityNextRunName ON jobs ( state, priority, nextRun, name );
```

Then copied the subquery to a script that ran the query 1000 times

I ran this twice for each version of the query.
`ORDER BY nextRun ASC` had an average runtime of `0.003772` and `0.003973`
`ORDER BY nextRun ASC, jobID ASC` had an average runtime of `0.004018` and `0.003936`

I think the difference is inconsequential and I think we should move forward with these changes

File: plugins/Jobs.cpp [hunk 16]
```diff
@@ -442,7 +465,7 @@ bool BedrockPlugin_Jobs::processCommand(SQLite& db, BedrockCommand& command) {
             int64_t parentJobID = SContains(job, "parentJobID") ? SToInt(job["parentJobID"]) : 0;
             if (parentJobID) {
                 SQResult result;
-                if (!db.read("SELECT state, parentJobID FROM jobs WHERE jobID=" + SQ(parentJobID) + ";", result)) {
+                if (!db.read("SELECT state, parentJobID, retryAfter FROM jobs WHERE jobID=" + SQ(parentJobID) + ";", result)) {
```
Comment (mea36) [inline plugins/Jobs.cpp]: I'm not sure why this is here, removing it.

File: plugins/Jobs.cpp (hunk 15 above)
[the remark this answers is not in the document]
Comment (cead22) [reply plugins/Jobs.cpp]: 👍

File: plugins/Jobs.cpp (hunk 16 above)
Comment (mea36) [reply plugins/Jobs.cpp]: This is here because auto-retrying parents shouldn't have children - https://github.com/Expensify/Bedrock/pull/169/files#diff-d116ebb4bb9e2a3fd799045c7b182707R372

Adding this back.  I think it disappeared in a bad merge with master when we added `parentJobID` to this query

Comment (mea36) [reply plugins/Jobs.cpp]: Never mind, I'm confusing myself.  We check this in `peek` and the code I linked to is alerting, not throwing.  Removing this

**Comment by mea36** [comment]
Woot! Tests are passing and I think I addressed everyone's comments

File: test/tests/jobs/GetJobTest.cpp [hunk 17]
```diff
@@ -0,0 +1,408 @@
...
+        command["query"] = "DELETE FROM jobs WHERE jobID > 0;";
+        tester->executeWaitVerifyContent(command);
+    }
+
+    void tearDownClass() { delete tester; }
+
+    // Simple GetJob
+    void getJob() {
+        // Create the job
+        SData command("CreateJob");
```
Comment (cead22) [inline test/tests/jobs/GetJobTest.cpp]: You could DRY this file with a func to create a job that takes name, priority, nextRun, etc. Nab

File: test/tests/jobs/GetJobTest.cpp [hunk 18]
```diff
@@ -0,0 +1,408 @@
...
+
+        // Medium
+        command["name"] = "medium_4";
+        command["priority"] = "500";
+        response = tester->executeWaitVerifyContentTable(command);
+        jobList.push_back(response["jobID"]);
+
+        // Confirm these jobs all have the same nextRun time
+        SQResult result;
+        tester->readDB("SELECT nextRun FROM jobs WHERE jobID IN (" + SComposeList(jobList) + ") GROUP BY nextRun;", result);
```
Comment (cead22) [inline test/tests/jobs/GetJobTest.cpp]: SELECT DISTINTC is clearer imo, nab

File: test/tests/jobs/GetJobTest.cpp [hunk 19]
```diff
@@ -0,0 +1,408 @@
...
+        command["priority"] = "0";
+        command["firstRun"] = SComposeTime("%Y-%m-%d %H:%M:%S", STimeNow());
+        STable response = tester->executeWaitVerifyContentTable(command);
+        list<string> jobList;
+        jobList.push_back(response["jobID"]);
+
+        // Medium
+        command["name"] = "medium";
+        command["priority"] = "500";
+        command["firstRun"] = SComposeTime("%Y-%m-%d %H:%M:%S", STimeNow()+1000000);
```
Comment (cead22) [inline test/tests/jobs/GetJobTest.cpp]: You should use spaces around `+`

File: test/tests/jobs/GetJobTest.cpp [hunk 20]
```diff
@@ -0,0 +1,408 @@
...
+        ASSERT_EQUAL(response["name"], "low");
+        sleep(1);
+        response = tester->executeWaitVerifyContentTable(command);
+        ASSERT_EQUAL(response["name"], "medium");
+        sleep(1);
+        response = tester->executeWaitVerifyContentTable(command);
+        ASSERT_EQUAL(response["name"], "high");
+
+        // Sleep 3 seconds so that the nextRun time for the last 'high' job is now
+        sleep(3);
```
Comment (cead22) [inline test/tests/jobs/GetJobTest.cpp]: Either remove this and change the assertions below or change the comment in the description ;)
`// Expect the jobs to be returned in order of low, medium, high, low, medium, high`

File: test/tests/jobs/GetJobTest.cpp [hunk 21]
```diff
@@ -0,0 +1,408 @@
...
+        // Checking the cancelled job
+        list<string> cancelledChildJobs = SParseJSONArray(response["cancelledChildJobs"]);
+        ASSERT_EQUAL(cancelledChildJobs.size(), 1);
+        childJob = SParseJSONObject(cancelledChildJobs.front());
+        ASSERT_EQUAL(childJob["jobID"], cancelledChildID);
+        ASSERT_EQUAL(childJob["data"], cancelledChildData);
+    }
+
+    // This is the same as testPriorities but some of the states are set to RUNQUEUED
+    // So we set the the firstRun for all the jobs that don't have a retryAfter to 1 second in the future
```
Comment (cead22) [inline test/tests/jobs/GetJobTest.cpp]: `the the`

File: test/tests/jobs/GetJobTest.cpp [hunk 22]
```diff
@@ -0,0 +1,408 @@
...
+        // GetJob and confirm that the last 3 jobs are returned in priority order since now is past nextRun for all of them
+        response = tester->executeWaitVerifyContentTable(command);
+        ASSERT_EQUAL(response["name"], "high");
+        response = tester->executeWaitVerifyContentTable(command);
+        ASSERT_EQUAL(response["name"], "medium");
+        response = tester->executeWaitVerifyContentTable(command);
+        ASSERT_EQUAL(response["name"], "low");
+    }
+
+    // get a parent job that has finished and cancelled jobs
```
Comment (cead22) [inline test/tests/jobs/GetJobTest.cpp]: `Get`

**Comment by mea36** [comment]
Updated

**Comment by iwiznia** [comment]
This seems to be ready and all comments from reviews where addressed. Asking @pecanoro to merge so I can finish SSDos. If there's anything missing, we can address it on a different PR

**Comment by fnwbr** [comment]
Apart from the included tests there are no further QA steps necessary?

**Comment by mea36** [comment]
I think the tests should cover it

File: plugins/Jobs.cpp (hunk 15 above)
[the remark this answers is not in the document]
Comment (quinthar) [reply plugins/Jobs.cpp]: Looks like our performance testing wasn't adequate here -- when harvesting started, 1.5M jobs became runnable, and performance slowed to a near-halt.  What's interesting is we had more or less this exact fear, and actually thought we tested to confirm it, but still missed it.  What can we do differently next time to catch this?  Only do performance testing on prod, and only right after harvesting starts?

[the remark this answers is not in the document]
Comment (quinthar) [reply plugins/Jobs.cpp]: Cc: @coleaeason

[the remark this answers is not in the document]
Comment (coleaeason) [reply plugins/Jobs.cpp]: Well one way to catch this next time would be to make sure we time queries while harvesting is running and while it's not running. 

Another way to catch this would be if we could build a test that mimics harvesting, i.e. it queues 1.5mn run ready jobs that have similar parameters to harvesting jobs (retry times, data, etc.). It would be nice if we had a way to fail a test if a query or operation takes more time than X.

[the remark this answers is not in the document]
Comment (iwiznia) [reply plugins/Jobs.cpp]: > Well one way to catch this next time would be to make sure we time queries while harvesting is running and while it's not running.

I agree, running queries in the jobs table when the queue is almost empty is almost as running them on dev. 
Not sure how big is the jobs table, but maybe we can grab a snapshot of it when it has data and save it to use it in any performance test we need.
